sdk/communication/communication-common/src/entraTokenCredential.ts

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { AccessToken } from "@azure/core-auth";
+import { TokenCredential } from "@azure/identity";
+import {
+  TokenCredential as AcsTokenCredential,
+  CommunicationGetTokenOptions,
+} from "./communicationTokenCredential";
+import { AbortSignalLike } from "@azure/abort-controller";
+import { Client, getClient } from "@azure-rest/core-client";
+import {
+  HttpClient,
+  createDefaultHttpClient,
+  createHttpHeaders,
+  createPipelineRequest,
+} from "@azure/core-rest-pipeline";
+
+export interface ExchangeTokenResponse {
+  identity: string;
+  accessToken: {
+    token: string;
+    expiresOn: string;
+  };
+}
+
+export interface EntraCommunicationTokenCredentialOptions {
+  /**
+   * The Azure Communication Service resource endpoint URL, e.g. https://myResource.communication.azure.com.
+   */
+  resourceEndpoint: string;
+  /**
+   * The Entra ID token credential.
+   */
+  tokenCredential: TokenCredential;
+  /**
+   * The scopes for retrieving the Entra ID access token.
+   */
+  scopes: string[];
+}
+
+/**
+ * EntraTokenCredential
+ */
+export class EntraTokenCredential implements AcsTokenCredential {
+  private isPending: Promise<AccessToken> | null;
+  private result = {
+    entraToken: undefined as string | undefined,
+    acsToken: { token: "", expiresOnTimestamp: 0 },
+  };
+  private client: Client;
+  private httpClient: HttpClient;
+
+  constructor(private options: EntraCommunicationTokenCredentialOptions) {
+    this.client = getClient(options.resourceEndpoint);
+    this.httpClient = createDefaultHttpClient();
+
+    // immediately fetch the token to pre-warm
+    this.isPending = this.getToken();
+  }
+
+  public async getToken(options?: CommunicationGetTokenOptions): Promise<AccessToken> {
+    if (options?.abortSignal?.aborted) {
+      return { token: "", expiresOnTimestamp: 0 };
+    }
+
+    // we're awaiting the token fetch, so we don't want to start another one
+    // however, we're ignoring the new abortSignal, unfortunately
+    if (!this.isPending) {
+      this.isPending = this.getTokenInternal(options);
+    }
+
+    try {
+      await this.isPending;
+    } finally {
+      this.isPending = null;
+    }
+
+    return this.result.acsToken;
+  }
+
+  private async getTokenInternal(options?: CommunicationGetTokenOptions): Promise<AccessToken> {
+    const getTokenOptions = options?.abortSignal ? { abortSignal: options.abortSignal } : undefined;
+    const token = await this.options.tokenCredential.getToken(this.options.scopes, getTokenOptions);
+
+    if (token === null) {
+      this.result = {
+        entraToken: undefined,
+        acsToken: { token: "", expiresOnTimestamp: 0 },
+      };
+    } else if (this.result.acsToken.token === "" || token.token !== this.result.entraToken) {
```

Comment:
I think here we have to check if we have valid acs token cashed? If yes then fetch entra token after, not before checking the cache

Reply:
Calling tokenCredential.getToken always at the beginning should be okay because it's caching internally.
I think the implicit assumption from me here was that if the ACS token is expired, we must also have gotten a new Entra token and the second clause would be true. But it's better to add an explicit OR here to check if the ACS token has expired.